### DISCUSSION TITLE
Harmony batch correction for integration of snATAC-seq and scRNA-seq data

### DISCUSSION THREAD

**Discussion by jacklyn-ashmus** [satijalab/seurat#4423] [Q&A] [answered]
Hi,
I've been having some issues trying to get Harmony to work correctly using Seurat. I can get it to batch correct our RNA data, but when I use Harmony through Seurat, I somehow can't get it to correct the batch effects at all. The resulting atac-seq UMAP plots look kind of like spaghetti and meatballs. This problem persists even after I transfer anchors and merge the two data sets, and then call Harmony. I have tried both 
`hm_atac <- HarmonyMatrix(Embeddings(seurat_ATAC, reduction='lsi'), seurat_ATAC$library, do_pca=FALSE)`
and 
`atac_adata <- RunHarmony(atac_adata, group.by.vars='library', reduction.save='harmony')`
after calling normalization and PCA, and after making sure that the correct variables are in the 'library' metadata column and that I have the correct assay active. 

I should probably also mention that I've successfully run Harmony on this ATAC data set outside of the context of Seurat before, where the only difference in commands was to include the parameter tau=2. I tried that here and it didn't make a difference.

I've included my code and some figures in this post, although I'm fairly new to github, so I apologize if anything is wrong with this.

Edit: 
OK, here are the fragments of code that I have been using that I think are involved in the issue:

reading in the h5ad files (done within a loop, for every sample):
```
atac_seurat_list <- list()

loop:

    h5FileName <- sprintf('/dev/ATAC_MTX/%s.raw_peak_bc_matrix.h5', sample)
    h5 <- Read10X_h5(h5FileName, use.names = TRUE, unique.features = TRUE)
    pbmc_adata_atac_single <- CreateSeuratObject(counts = h5, assay = 'ATAC', min.cells = 3, min.features = 200)
    pbmc_adata_atac_single[['percent.mt']] <- PercentageFeatureSet(pbmc_adata_atac_single, pattern = '^MT-')
    DefaultAssay(pbmc_adata_atac_single) <- 'ATAC'
    atac_seurat_list[[sample]]<- pbmc_adata_atac_single
```

merging the obects together: 
```
mergeLeftList <- c(atac_seurat_list$pbmc2, atac_seurat_list$pbmc3, atac_seurat_list$pbmc4, 
                  atac_seurat_list$pbmc5, atac_seurat_list$pbmc6, atac_seurat_list$pbmc8, 
                   atac_seurat_list$pbmc9, atac_seurat_list$pbmc10, atac_seurat_list$pbmc12)
atac_adata <- merge(
  x = atac_seurat_list$pbmc1,
  y = mergeLeftList
) 

```

normalization and running Harmony, etc:
```
atac_adata <- RunTFIDF(atac_adata)
atac_adata <- FindTopFeatures(atac_adata, min.cutoff = "q0")
atac_adata <- RunSVD(atac_adata)
atac_adata <- RunHarmony(atac_adata, group.by.vars='library', reduction.save = "harmony", assay.use="ATAC")
atac_adata <- RunUMAP(atac_adata, reduction = "lsi", dims = 2:30, reduction.name = "umap.atac", reduction.key='atacUMAP_')
atac_adata <- FindNeighbors(object = atac_adata, reduction = 'harmony', dims = 2:30)
 atac_adata <- FindClusters(object = atac_adata, algorithm=4, verbose=FALSE, method="igraph", resolution=0.5)
```

![ATAC-only-harmony](https://user-images.githubusercontent.com/22038654/116720124-b3b06a00-a990-11eb-9e40-a0178478f630.png)
![ATAC-only-umap](https://user-images.githubusercontent.com/22038654/116720128-b4490080-a990-11eb-96a6-82bc305b5bd6.png)
![coembed-samples-harmony](https://user-images.githubusercontent.com/22038654/116720130-b4e19700-a990-11eb-9d46-5fbbe176b7f9.png)
![coembed-tech-harmony](https://user-images.githubusercontent.com/22038654/116720134-b57a2d80-a990-11eb-874b-1d9ab430630a.png)
![coembed-samples-UMAP](https://user-images.githubusercontent.com/22038654/116720136-b612c400-a990-11eb-9ce4-9350eb1b87bb.png)
![coembed-tech-UMAP](https://user-images.githubusercontent.com/22038654/116720138-b612c400-a990-11eb-9856-e756bd12d8a6.png)

**Comment by timoast** [comment, marked as answer]
Hi, could you provide the minimal code required to reproduce the issue you're having, and put it in the post itself rather than a pdf?

**Reply by jacklyn-ashmus** [reply]
Hi, I am really sorry about that. I've fixed it to only include the relevant bits of code now.

**Reply by timoast** [reply]
I would recommend quantifying the same peak set across the different samples you have before merging them. See the examples given here: https://satijalab.org/signac/articles/merging.html
